### PR TITLE
Add Hybrid Engine rollout in DeepSpeed to support On-Policy Distillation (OPSD) Trainer

### PR DESCRIPTION
**Pull request by PKUWZP**
## Summary

This PR adds a DeepSpeed-native on-policy distillation trainer. It also incorporates the PR from @delock (https://github.com/PKUWZP/DeepSpeed/pull/1), which abstracts the OPSD example into DeepSpeed submodules, including a rollout engine abstract layer, two rollout implementations (HybridEngine and vLLM), and OPSD trainer.

On-policy distillation: a small **student** generates rollouts, a frozen large **teacher** scores them, and the student is updated by a per-token divergence (forward-KL / reverse-KL / JSD) between the two distributions on the student's own samples. Each step has three phases — student rollout → teacher forward + CPU logit cache → student forward + streamed divergence + backward — so the full `[B, T, V]` teacher tensor never co-resides with the student logits on the training device.

## Key Design Decisions

- Rollout engine abstracted as RolloutEngine ABC, created via `build_rollout()` factory
- `HybridEngineRollout` runs in-process, reusing model weights — no cross-process weight transfer needed
- `VLLMRollout` runs as a subprocess to avoid vLLM's `new_group()` deadlocking with DeepSpeed launcher
- `HybridEngineRollout` support continuous batching and graph capture

**Modules**:

- `deepspeed/runtime/rlhf/trainer/opsd.py` — chunked / streamed forward-KL, reverse-KL, JSD with sequence-axis chunking
- `deepspeed/runtime/rlhf/trainer/teacher.py` — frozen teacher wrapper + `TeacherLogitCache` (host-resident, chunk fetch)
- `deepspeed/runtime/rlhf/rollout/hybrid_engine_rollout.py` — HybridEngine rollout backends
- `deepspeed/runtime/rlhf/rollout/vllm_rollout.py` — vLLM rollout backends
- `benchmarks/opsd` — Benchmark scripts for OPSD training 
- `tests/unit/runtime/rollout` — testing scripts for OPSD rollout backends (vLLM + HybridEngine) 

**Validated end-to-end**: On 2× H200 with Qwen2.5-0.5B-Instruct student + Qwen2.5-1.5B-Instruct teacher via the hybrid-engine path; loss finite for 5 steps. See README for the smoke recipe.

**Follow-up items**: documented in README file and SGLang rollout integration. 

## Test plan

- [x] `cd examples/opsd && python -m pytest tests/ -v` → **87/87 passing** on CPU
- [x] `deepspeed --num_gpus 2 main.py --config configs/smoke_hybrid.json` end-to-end on 2× H200 → 5 finite-loss steps
- [x] `pre-commit run --files <all changed files>` → green (yapf, flake8, check-torchdist, check-license, check-torchcuda, codespell)
- [x] vLLM rollout end-to-end
- [ ] Larger-scale training run (out of scope for the initial PR)

### CODE REVIEW

**Comment by sfc-gh-truwase** [comment]
@PKUWZP, this needs to be split between examples and engine logic. 

The examples should go to the examples repo: https://github.com/deepspeedai/DeepSpeedexamples

The engine should go somewhere under [deepspeed](https://github.com/deepspeedai/DeepSpeed/tree/master/deepspeed). Since this already an existing [hybrid_engine](https://github.com/deepspeedai/DeepSpeed/blob/master/deepspeed/runtime/hybrid_engine.py), we should figure out if it makes sense to replace or consolidate.

File: benchmarks/opsd/bench_14b_rollout.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (sfc-gh-truwase) [inline benchmarks/opsd/bench_14b_rollout.py]: Please move benchmarks to examples repo. That will reduce the dependencies for building DS as benchmarks increase in complexity.

File: deepspeed/runtime/rollout/vllm_rollout.py [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+
+logger = logging.getLogger(__name__)
+
+_HTTP_TIMEOUT = 120
+_VLLM_NCCL_BACKEND = "nccl"
+
+
+def _gdr_available() -> bool:
+    try:
+        return torch.cuda.is_available() and torch.cuda.nccl.version() is not None  #ignore-cuda
```
Comment (sfc-gh-truwase) [inline deepspeed/runtime/rollout/vllm_rollout.py]: These `cuda` references is breaking accelerator abstraction.

File: deepspeed/runtime/rollout/vllm_rollout.py [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+        }
+
+        init_thread = threading.Thread(target=self._post,
+                                       args=("/init_weight_transfer_engine", ),
+                                       kwargs={"json": {
+                                           "init_info": init_info
+                                       }})
+        init_thread.start()
+
+        from vllm.distributed.utils import StatelessProcessGroup
```
Comment (sfc-gh-truwase) [inline deepspeed/runtime/rollout/vllm_rollout.py]: Why are we making vllm a dependency? This breaks the lightweight nature of DS and introduces unnecessary complexity.

**Comment by delock** [comment]
As discussed, this PR would go to the following direction:
1. Add Hybrid Engine rollout in DeepSpeed, in order to support OPSD trainer.
2. OPSD trainer will be added to DeepSpeedExamples as a seperate app
3. Hybrid Engine rollout will provided colocated rollout to simplify OPSD training process.

**Review by delock — APPROVED** [review]
